resources-service/src/resource/resource-model.ts

```diff
@@ -15,9 +15,62 @@
  */
 
 import { AccountSID } from '@tech-matters/twilio-worker-auth';
-import { getById, ReferreableResource } from './resource-data-access';
+import {
+  getById,
+  getByIdList,
+  getWhereNameContains,
+  ReferrableResource,
+} from './resource-data-access';
+
+export type SearchParameters = {
+  ids: string[];
+  nameSubstring?: string;
+  pagination: {
+    limit: number;
+    start: number;
+  };
+};
+
+const EMPTY_RESULT = { totalCount: 0, results: [] };
+const MAX_SEARCH_RESULTS = 200;
 
 export const getResource = (
   accountSid: AccountSID,
   resourceId: string,
-): Promise<ReferreableResource | null> => getById(accountSid, resourceId);
+): Promise<ReferrableResource | null> => getById(accountSid, resourceId);
+
+export const searchResources = async (
+  accountSid: AccountSID,
+  { nameSubstring, ids = [], pagination: { limit: unboundedLimit, start } }: SearchParameters,
+): Promise<{ totalCount: number; results: ReferrableResource[] }> => {
+  const limit = Math.min(MAX_SEARCH_RESULTS, unboundedLimit);
+  const { results: idsOfNameMatches, totalCount: nameSearchTotalCount } = nameSubstring
+    ? await getWhereNameContains(accountSid, nameSubstring, start, limit)
+    : EMPTY_RESULT;
+
+  // If I'd known this logic would be such a hairball when I started writing the tests I wouldn't have bothered
+  // Still, it is only temporary, once the real search is implemented we won't need to splice 2 sets of results together so it will be much simpler
+  const idsToLoad = [...idsOfNameMatches, ...ids.filter(id => !idsOfNameMatches.includes(id))];
+  if (!idsToLoad.length) return { results: [], totalCount: nameSearchTotalCount };
+  // This might well be more than needed to meet the 'limit' criteria but best to query them all in case any of them 'miss'
+  const unsortedResourceList = await getByIdList(accountSid, idsToLoad);
+  const resourceMap = Object.fromEntries(
+    unsortedResourceList.map(resource => [resource.id, resource]),
+  );
+
+  // Add ALL the resources found looking up specific IDs to the paginated block of name search results
+  const untrimmedResults = idsToLoad
+    .map(id => {
+      const mappedValue = resourceMap[id];
+      // So each value is only used once
+      delete resourceMap[id];
+      return mappedValue;
+    })
+    .filter(r => r);
```

Comment:
Is this deduplicating entries? If so, would it be a good idea dedup the ids before actually getting them like
![Screenshot from 2023-02-07 11-01-20](https://user-images.githubusercontent.com/15805319/217266374-0b8144dc-d06f-4e9c-b234-f340773746a9.png)
?

Reply:
I know this is temporary so it may not be worth. The "temporary" means that this will be replaced with the AWS search engine at some point?

Reply:
Sets are unordered collections, so it's implementation dependent whether it preserves the original order when you create a set from an array then iterate over the values. This approach is guaranteed to preserve the original order of the IDs (discarding all but the first in the case of duplicates)

Since the duplicated IDs won't return duplicated rows from the DB if you query with duplicated IDs, it doesn't make much difference whether you drop the duplicates before or after the query

Reply:
And the order here matters because that will be a "weighted sorted list" in the future, when we actually hit the search indexes, right?